source/timed-out.js

```diff
@@ -0,0 +1,78 @@
+'use strict';
+
+// Forked from https://github.com/floatdrop/timed-out
+
+module.exports = function (req, time) {
+	if (req.timeoutTimer) {
+		return req;
+	}
+
+	const delays = isNaN(time) ? time : {socket: time, connect: time};
+	const host = req._headers ? (' to ' + req._headers.host) : '';
+
+	function throwESOCKETTIMEDOUT() {
+		req.abort();
+		const e = new Error('Socket timed out on request' + host);
+		e.code = 'ESOCKETTIMEDOUT';
+		req.emit('error', e);
+	}
+
+	function throwETIMEDOUT() {
+		req.abort();
+		const e = new Error('Connection timed out on request' + host);
+		e.code = 'ETIMEDOUT';
+		req.emit('error', e);
+	}
+
+	if (delays.connect !== undefined) {
+		req.timeoutTimer = setTimeout(throwETIMEDOUT, delays.connect);
+	}
+
+	if (delays.request !== undefined) {
+		req.requestTimeoutTimer = setTimeout(() => {
+			clear();
+
+			if (req.connection.connecting) {
+				throwESOCKETTIMEDOUT();
+			} else {
+				throwETIMEDOUT();
+			}
+		}, delays.request);
+	}
+
+	// Clear the connection timeout timer once a socket is assigned to the
+	// request and is connected.
+	req.on('socket', socket => {
+		// Socket may come from Agent pool and may be already connected.
+		if (!(socket.connecting || socket._connecting)) {
+			connect();
+			return;
+		}
+
+		socket.once('connect', connect);
+	});
+
+	function clear() {
+		if (req.timeoutTimer) {
+			clearTimeout(req.timeoutTimer);
+			req.timeoutTimer = null;
+		}
+	}
+
+	function connect() {
+		clear();
+
+		if (delays.socket !== undefined) {
+			// Abort the request if there is no activity on the socket for more
+			// than `delays.socket` milliseconds.
+			req.setTimeout(delays.socket, throwESOCKETTIMEDOUT);
+		}
+
+		req.on('end', () => {
```

Comment:
I don't think this is right. I think the `request` timeout was introduced in https://github.com/sindresorhus/got/pull/308, which describes it as the time for the entire response to be received.

`'end'` is not an event for `http.ClientRequest`, is it? It looks like tests around this are contrived to pass - I don't see a test that shows the timeout is not raised if the request (and response) completes before the timeout.

I think this should be something like:

```js
req.on('response', res => {
  res.on('end', () => {
    clearTimeout(req.requestTimeoutTimeout)
  })
})

Reply:
Fixin' right now. You have really sharp eyes!

Reply:
> It looks like tests around this are contrived to pass

Make another test and increase the timeout so the request finishes. Why hadn't you checked that? It passes. It doesn't throw, because the promise has been resolved already.

Reply:
> `await delay(reqDelay * 3);`

If you used a Promise and attached some listeners, it would throw with that too.

The point is no one has added a delay because everybody assumed if the request finishes without any errors, it won't throw then. That's just an oversight. If you tell that I just did this intentionally, then you're wrong.

Reply:
@jstewmon Could you submit the above test?

Reply:
Tests that have setup are always contrived to pass - I should have just said that tests are missing for cases where the timeout is not breached. I didn't mean that you deliberately introduced a bug. :-)